### PR TITLE
Label filter widget in mgmt console

### PR DESCRIPTION
**Pull request by jwforres**

### CODE REVIEW

**Comment by jwforres** [comment]
@liggitt

File: assets/app/scripts/services/labelFilter.js [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+
+angular.module('openshiftConsole')
+.factory('LabelFilter', [function() {
+  function LabelFilter() {
+    this._existingLabels = {};
+    this._activeFilters = {};
+    this._onActiveFiltersChangedCallbacks = $.Callbacks();
+  }
+
+  LabelFilter.prototype.createLabelSuggestionsFromResources = function(items, map) {
```
Comment (liggitt) [inline assets/app/scripts/services/labelFilter.js]: Call this `addLabelSuggestionsFromResources`

File: assets/app/scripts/controllers/builds.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      $scope.$apply(function() {
+        $scope.unfilteredBuilds = builds.by("metadata.name");
+        LabelFilter.createLabelSuggestionsFromResources($scope.unfilteredBuilds, $scope.labelSuggestions);
+        LabelFilter.setLabelSuggestions($scope.labelSuggestions);
+        $scope.builds = LabelFilter.filterResources($scope.unfilteredBuilds);
+      });
+
+      console.log("builds (subscribe)", $scope.unfilteredBuilds);
+    };
+    $scope.watches.push(DataService.watch("builds", $scope, buildsCallback));
```
Comment (liggitt) [inline assets/app/scripts/controllers/builds.js]: are we counting on the project controller to clean up our watches?

File: assets/app/scripts/services/labelFilter.js [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+
+  LabelFilter.prototype.isResourceIncludedInActiveFilters = function(resource) {
+    if (!resource) {
+      return false;
+    }
+    var labels = resource.labels || {};
+    if (resource.metadata) {
+      labels = resource.metadata.labels || {};
+    }
+    for (var id in this._activeFilters) {
```
Comment (liggitt) [inline assets/app/scripts/services/labelFilter.js]: Separate out the object that represents a label selector, and the logic to decide whether that matches a set of labels. Shouldn't be tied so closely to a LabelFilter widget (we'll want to be able to parse a label selector from an API object later). A LabelFilter widget should be able to return a LabelSelector object representing its active filters. Then `LabelSelector.select(resources)` can select resources (probably calling `LabelSelector.matches(resource)` internally)

File: assets/app/scripts/services/labelFilter.js [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+              if (labels[filter.key] == filter.values[i]) {
+                keep = true;
+              }
+            }
+          }
+          if (!keep) {
+            return false;
+          }
+          break;
+        case "NOT_IN":
```
Comment (liggitt) [inline assets/app/scripts/services/labelFilter.js]: can we match the actual syntax of label selectors?

File: assets/app/scripts/services/labelFilter.js [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+          var keep = true;
+          if (labels[filter.key]) {
+            for (var i = 0; i < filter.values.length; i++) {
+              if (labels[filter.key] == filter.values[i]) {
```
Comment (liggitt) [inline assets/app/scripts/services/labelFilter.js]: `keep = labels[filter.key] != filter.values[i]`

File: assets/app/scripts/services/labelFilter.js [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+            }
+          }
+          if (!keep) {
+            return false;
+          }
+          break;
+        case "NOT_IN":
+          var keep = true;
+          if (labels[filter.key]) {
+            for (var i = 0; i < filter.values.length; i++) {
```
Comment (liggitt) [inline assets/app/scripts/services/labelFilter.js]: `for (var i = 0; keep && i < filter.values.length; i++) {`

File: assets/app/scripts/services/labelFilter.js (hunk 3 above)
Comment (liggitt) [inline assets/app/scripts/services/labelFilter.js]: That'll also let us write tests around label selector matching (which we need)

File: assets/app/scripts/services/labelFilter.js [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+    var self = this;
+
+    $('.label-filter-key').selectize({
+      valueField: "key",
+      labelField: "key",
+      searchField: ["key"],
+      create: true,
+      persist: true, // i want this to be false but there appears to be a bug in selectize where setting
+                     // this to false has a side effect of causing items that were not created by the user
+                     // to also disappear from the list after being removed
```
Comment (liggitt) [inline assets/app/scripts/services/labelFilter.js]: upstream issue / patch?

File: assets/app/scripts/services/labelFilter.js [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+      preload: true,
+      onItemAdd: function(value, $item) {
+        var selectizeValues = $('.label-filter-values')[0].selectize;
```
Comment (liggitt) [inline assets/app/scripts/services/labelFilter.js]: is there a way to get this from the context (`this`, etc). class queries will break if we ever have more than one in a page (even if they're hidden)

File: assets/app/scripts/services/labelFilter.js [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+        selectizeValues.clearOptions();
+        selectizeValues.load(function(callback) {
+          var options = [];
+          var key = $('.label-filter-key')[0].selectize.getValue();
```
Comment (liggitt) [inline assets/app/scripts/services/labelFilter.js]: `$(...)[0]` is a code smell... any other way to get these?

File: assets/app/scripts/services/labelFilter.js [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+      .prop("filter-label-id", this._getIdForFilter(filter))
+      .click($.proxy(this, '_removeActiveFilter'))
+      .append(
+        $('<span>')
+          .text(this._getLabelForFilter(filter))
+          // TODO move to the less styles instead
+          .css("padding-right", "5px")
+      )
+      .append(
+        $('<i>')
```
Comment (liggitt) [inline assets/app/scripts/services/labelFilter.js]: italic?

File: assets/app/views/builds.html [hunk 11]
```diff
@@ -1,20 +1,34 @@
...
-    <div style="margin-bottom: 10px;" ng-repeat="build in builds">
-      <h3>{{build.metadata.labels.buildconfig}} <span class="small">({{build.metadata.name}})</span></h3>
-      <div>Created: <relative-timestamp timestamp="build.metadata.creationTimestamp"></relative-timestamp></div>            
-      <div>Status: {{build.status}}</div>
-      <div>Build strategy: {{build.parameters.strategy.type}}</div>
-      <div ng-if="build.parameters.strategy.type == 'STI'">
-        Builder image: {{build.parameters.strategy.stiStrategy.builderImage}}
+    <div ng-controller="BuildsController">
+      <h1>Builds</h1>
+      <div ng-if="(builds | hashSize) == 0">
```
Comment (liggitt) [inline assets/app/views/builds.html]: Can we move this logic into the controllers? It seems better to have the controller indicate different states:
```
info, "Loading..."
info, "No builds to show"
warning, "Active filters are hiding all builds"
error, "Could not retrieve builds"
```
Then the view can just render the empty message and status

File: assets/app/scripts/services/labelFilter.js (hunk 10 above)
Comment (jwforres) [inline assets/app/scripts/services/labelFilter.js]: see the font-awesome examples http://fortawesome.github.io/Font-Awesome/examples/

not necessary to be an <i> but a common shorthand

File: assets/app/views/builds.html (hunk 11 above)
Comment (jwforres) [inline assets/app/views/builds.html]: Yeah then I could actually make the template rendering a separate directive

File: assets/app/scripts/services/labelFilter.js (hunk 1 above)
Comment (jwforres) [inline assets/app/scripts/services/labelFilter.js]: Fixed

File: assets/app/scripts/controllers/builds.js (hunk 2 above)
Comment (jwforres) [inline assets/app/scripts/controllers/builds.js]: Fixed, individual controllers handle their own watches.

File: assets/app/scripts/services/labelFilter.js (hunk 4 above)
Comment (jwforres) [inline assets/app/scripts/services/labelFilter.js]: Fixed, as best as possible.

File: assets/app/scripts/services/labelFilter.js (hunk 5 above)
Comment (jwforres) [inline assets/app/scripts/services/labelFilter.js]: Fixed.

File: assets/app/scripts/services/labelFilter.js (hunk 6 above)
Comment (jwforres) [inline assets/app/scripts/services/labelFilter.js]: Fixed.

File: assets/app/scripts/services/labelFilter.js (hunk 7 above)
Comment (jwforres) [inline assets/app/scripts/services/labelFilter.js]: No, the behavior is acceptable for now as we discussed. If it gets annoying we can look at patching upstream.

File: assets/app/scripts/services/labelFilter.js (hunk 8 above)
Comment (jwforres) [inline assets/app/scripts/services/labelFilter.js]: Fixed, lots of changes around this.

File: assets/app/scripts/services/labelFilter.js (hunk 9 above)
Comment (jwforres) [inline assets/app/scripts/services/labelFilter.js]: Fixed in refactor to stop using css selectors.

File: assets/app/views/builds.html (hunk 11 above)
Comment (jwforres) [inline assets/app/views/builds.html]: Refactored alerts into a directive.  Added emptyMessage set by the controllers (goes from 'Loading...' to 'No whatevers to show')

File: assets/app/scripts/services/labelFilter.js (hunk 3 above)
Comment (jwforres) [inline assets/app/scripts/services/labelFilter.js]: Fixed in latest

**Comment by jwforres** [comment]
rebased and squashed

**Comment by liggitt** [comment]
LGTM, merge at will

**Comment by jwforres** [comment]
console_filter_active
console_filter_step1
console_filter_step2
console_filter_step3
console_navigation

**Comment by jwforres** [comment]
[merge]

**Pull request merged**